### PR TITLE
replace better version of roundpupil

### PR DESCRIPTION
**Pull request by johanmazoyer**
This PR is modifying the roundpupil function:
- remove the no_pixel keyword that was confusing
- create a grey_pup_bin_factor parameter. if grey_pup_bin_factor > it does a grey pupil, if grey_pup_bin_factor =1, normal one. Added Value errors to make it run smoothly

### CODE REVIEW

File: Asterix/optics/phase_amplitude_functions.py [hunk 1]
```diff
@@ -3,52 +3,77 @@
...
         Output circular pupil
     """
 
-    if no_pixel:
-        factor_bin = int(10)
-        pup_large = roundpupil(int(2 * prad) * factor_bin, factor_bin * prad, no_pixel=False)
-        return crop_or_pad_image(rebin(pup_large, factor=factor_bin, center_on_pixel=False), dim_pp)
+    if grey_pup_bin_factor > 1:
+        if grey_pup_bin_factor != int(grey_pup_bin_factor):
+            raise ValueError("grey_pup_bin_factor must be an integer")
```
Comment (ivalaginja) [inline Asterix/optics/phase_amplitude_functions.py:43]: For clarity in the raised exception and its message, you could add: "[...], currently it is {grey_pup_bin_factor}" and make it an f-string.

File: Asterix/tests/test_propagations.py [hunk 2]
```diff
@@ -12,9 +12,11 @@ def test_mft_centering():
 
     samp = 4
     efield = mft(pup, real_dim_input=pdim, dim_output=pdim, nbres=samp)
-    img = np.abs(efield) ** 2
+    img = np.abs(efield)**2
 
-    four_equal_pixels = (img[3,3] == img[3,4]) and (img[3,3] == img[4,3]) and (img[4,3] == img[4,3]) and (img[4,4] == img[4,4])
+    four_equal_pixels = (img[3, 3] == img[3, 4]) and (img[3, 3] == img[4, 3]) and (img[4, 3]
+                                                                                   == img[4, 3]) and (img[4, 4]
+                                                                                                      == img[4, 4])
```
Comment (ivalaginja) [inline Asterix/tests/test_propagations.py:19]: 😭 hard-lining on max line length is not always useful... :P

File: Asterix/tests/test_pupil.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 def test_roundpupil():
     dim = 10
     rad = int(dim/2)
-    pupil = roundpupil(dim, rad, no_pixel=False)
+    pupil = roundpupil(dim, rad, grey_pup_bin_factor = 10)
```
Comment (ivalaginja) [inline Asterix/tests/test_pupil.py:7]: No spaces around math operator in a function call.

**Review by ivalaginja — CHANGES_REQUESTED** [review]
Some minor comments to please look at, but overall this is compact and easy to merge.

File: Asterix/optics/phase_amplitude_functions.py [hunk 4]
```diff
@@ -3,52 +3,77 @@
...
     pupilnormal : 2D array
         Output circular pupil
     """
 
-    if no_pixel:
-        factor_bin = int(10)
-        pup_large = roundpupil(int(2 * prad) * factor_bin, factor_bin * prad, no_pixel=False)
-        return crop_or_pad_image(rebin(pup_large, factor=factor_bin, center_on_pixel=False), dim_pp)
+    if grey_pup_bin_factor > 1:
+        if grey_pup_bin_factor != int(grey_pup_bin_factor):
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/optics/phase_amplitude_functions.py:42]: you mean ?
`if not isinstance(grey_pup_bin_factor, int):`

[the remark this answers is not in the document]
Comment (ivalaginja) [reply Asterix/optics/phase_amplitude_functions.py]: yes of course, my mistake

**Pull request merged**